mftb.md

```diff
@@ -0,0 +1,351 @@
+# Mina Fungible _Tokenbase_[^tokenbase] Discussion
+
+This document is part specification and part meta-discussion. The specification portion touches on a
+design for a "Fungible Tokenbase", which outlines an approach to modeling fungible tokens on Mina.
+This includes interfaces for common actions, descriptions of how actions should be reduced into
+subsequent states, and the fungible token lifecycle. It describes the minimum functionality with
+which admin contracts can compose use-case-specific fungible tokens while still providing a common
+interface against which the community (namely wallets) can develop. The process of speccing this out
+led to critical questions about not only the possibility of implementation, but also the ultimate
+purpose of the specification itself.
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [Mina Fungible _Tokenbase_\[^tokenbase\] Discussion](#mina-fungible-tokenbasetokenbase-discussion)
+  - [Purpose of Specification](#purpose-of-specification)
+    - [Long-term Consideration](#long-term-consideration)
+  - [Near-term Recommendation](#near-term-recommendation)
+  - [Design](#design)
+    - [Goals](#goals)
+    - [Non-goals](#non-goals)
+    - [Actions](#actions)
+      - [`Create`](#create)
+      - [`Destroy`](#destroy)
+      - [`SetAdmin`](#setadmin)
+      - [`Mint`](#mint)
+      - [`Burn`](#burn)
+      - [`Transfer`](#transfer)
+      - [`Allocate`](#allocate)
+      - [`Deallocate`](#deallocate)
+      - [`Freeze`](#freeze)
+      - [`Thaw`](#thaw)
+      - [`SetMetadata`](#setmetadata)
+  - [Architecture](#architecture)
+    - [Actions and Reducers Drawbacks](#actions-and-reducers-drawbacks)
+  - [Closing Note](#closing-note)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Purpose of Specification
+
+Typically a specification would describe all the constraints that affect how one integrates with the
+specified software. In the context of Ethereum, [EIPs](https://eips.ethereum.org/) have been
+essential to the creation of a wide range of tools and a web of interconnected contracts. This has
+been possible because Ethereum contract execution happens on-chain. The creators of tools, dependent
+contracts, wallets and other such programs need only be aware of the types and capabilities of their
+target contract. **This is not true of Mina**. To interact with a Mina contract, one needs to run
+the contract code themselves. Each instruction affects how contract results are ultimately proven.
+Interacting with a contract whose method signatures abide by a given spec is not necessarily aligned
+with the desired behavior; in the world of Mina, the implementation is––more or less––the spec. This
```

Comment:
Spot-on. I agree with the stopgap approach you suggested of placing the contract behind a service. Otherwise, seemingly too much difficulty crops up for anyone who wants to interact with the contract. This difficulty is compounded as we get into large amounts of off-chain storage and upgradability.